src/main/java/core/basesyntax/MyLinkedList.java

```diff
@@ -3,45 +3,171 @@
+        head = null;
+        tail = null;
+        size = 0;
+    }
+
     @Override
     public void add(T value) {
+        addLast(value);
     }
 
     @Override
     public void add(T value, int index) {
+        if (index < 0 || index > size) {
+            throw new IndexOutOfBoundsException("Invalid index");
+        }
+        if (index == 0) {
+            addFirst(value);
+        } else if (index == size) {
+            addLast(value);
+        } else {
+            Node<T> current = getNode(index);
+            Node<T> newNode = new Node<>(value);
+            newNode.prev = current.prev;
+            newNode.next = current;
+            current.prev.next = newNode;
+            current.prev = newNode;
+            size++;
+        }
     }
 
     @Override
     public void addAll(List<T> list) {
+        for (T value : list) {
+            addLast(value);
+        }
     }
 
     @Override
     public T get(int index) {
-        return null;
+        indexPresent(index);
+        Node<T> current = getNode(index);
+        return current.data;
+    }
+
+    private Node<T> getNode(int index) {
+        indexPresent(index);
+        Node<T> current = head;
+        for (int i = 0; i < index; i++) {
+            current = current.next;
+        }
+        return current;
     }
 
     @Override
     public T set(T value, int index) {
-        return null;
+        indexPresent(index);
+        Node<T> current = getNode(index);
+        T oldValue = current.data;
+        current.data = value;
+        return oldValue;
     }
 
     @Override
     public T remove(int index) {
-        return null;
+        indexPresent(index);
+        Node<T> current = getNode(index);
+        if (size == 1) {
+            head = null;
+            tail = null;
+        } else if (current == head) {
+            head = head.next;
+            head.prev = null;
+        } else if (current == tail) {
+            tail = tail.prev;
+            tail.next = null;
+        } else {
+            current.prev.next = current.next;
+            current.next.prev = current.prev;
+        }
+        size--;
+        return current.data;
     }
 
     @Override
     public boolean remove(T object) {
+        Node<T> current = head;
+        while (current != null) {
+            if (current.data == object || (current.data != null && current.data.equals(object))) {
+                if (current == head) {
+                    head = head.next;
+                    if (head != null) {
+                        head.prev = null;
+                    }
+                } else if (current == tail) {
+                    tail = tail.prev;
+                    tail.next = null;
+                } else {
+                    current.prev.next = current.next;
+                    current.next.prev = current.prev;
+                }
+                size--;
+                return true;
+            }
+            current = current.next;
+        }
```

Comment:
methods have plenty of similar logic, introduce unlink(Node<T> node) method that will be called in both methods after required node found

Reply:
Done